src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,675 @@
+            self.de_init();
+        }
+
+        // TODO: Support keyboards other than the "virtual core keyboard device".
+        self.core_keyboard_id = (XKBXH.xkb_x11_get_core_keyboard_device_id)(self.xcb_connection);
+        let keymap = (XKBXH.xkb_x11_keymap_new_from_device)(
+            self.xkb_context,
+            self.xcb_connection,
+            self.core_keyboard_id,
+            xkbcommon_dl::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+        assert_ne!(keymap, ptr::null_mut());
+
+        let state = (XKBXH.xkb_x11_state_new_from_device)(
+            keymap,
+            self.xcb_connection,
+            self.core_keyboard_id,
+        );
+        self.post_init(state, keymap);
+    }
+
+    #[cfg(feature = "wayland")]
+    pub(crate) unsafe fn init_with_fd(&mut self, fd: File, size: usize) {
+        if !self.xkb_keymap.is_null() {
+            self.de_init();
+        }
+
+        let map = MmapOptions::new().len(size).map(&fd).unwrap();
+
+        let keymap = (XKBH.xkb_keymap_new_from_string)(
+            self.xkb_context,
+            map.as_ptr() as *const _,
+            ffi::xkb_keymap_format::XKB_KEYMAP_FORMAT_TEXT_V1,
+            ffi::xkb_keymap_compile_flags::XKB_KEYMAP_COMPILE_NO_FLAGS,
+        );
+
+        if keymap.is_null() {
+            panic!("Received invalid keymap from compositor.");
+        }
+
+        let state = (XKBH.xkb_state_new)(keymap);
+        self.post_init(state, keymap);
+    }
+}
+
+impl KbState {
+    #[cfg(feature = "wayland")]
+    pub(crate) unsafe fn key_repeats(&mut self, keycode: ffi::xkb_keycode_t) -> bool {
+        (XKBH.xkb_keymap_key_repeats)(self.xkb_keymap, keycode) == 1
+    }
+
+    #[inline]
+    pub(crate) fn ready(&self) -> bool {
+        !self.xkb_state.is_null()
+    }
+
+    #[inline]
+    #[cfg(feature = "wayland")]
+    pub(crate) fn locked(&self) -> bool {
+        self.locked
+    }
+
+    #[inline]
+    pub(crate) fn mods_state(&self) -> ModifiersState {
+        self.mods_state
+    }
+}
+
+impl Drop for KbState {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.xkb_compose_state.is_null() {
+                (XKBCH.xkb_compose_state_unref)(self.xkb_compose_state);
+            }
+            if !self.xkb_compose_state_2.is_null() {
+                (XKBCH.xkb_compose_state_unref)(self.xkb_compose_state_2);
+            }
+            if !self.xkb_compose_table.is_null() {
+                (XKBCH.xkb_compose_table_unref)(self.xkb_compose_table);
+            }
+            if !self.xkb_state.is_null() {
+                (XKBH.xkb_state_unref)(self.xkb_state);
+            }
+            if !self.xkb_keymap.is_null() {
+                (XKBH.xkb_keymap_unref)(self.xkb_keymap);
+            }
+            (XKBH.xkb_context_unref)(self.xkb_context);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    /// libxkbcommon is not available
+    XKBNotFound,
+}
+
+impl KbState {
```

Comment:
Merge other Kbd stuff, don't split impl like that. In general, just merge everything to the same `impl`. Use master layout for struct definitions. Applies to other code in linux backend, don't really care about the code style practices outside the linux backend.

```rust
struct Foo {
}

impl Foo {
  fn foo() -> OtherType {}
}

impl Drop for Foo{
}

impl OtherTrait for Foo {
}

struct OtherType {}
```

Reply:
Done in my PR.